README.md
# ***MrMineBOT***
A small program for the game, Mr. Mine, that lets you "sell all ores" and "forge gems" automatically.  

The program uses image recognition from *pyautogui* module to press certain buttons in a specific order. 

Image recognition is made for a *1920x1080 resolution display*. Other resolutions may not work - Haven't been tested with other resolutions.
Only tested on Windows.

---
This is just a small hobby project to keep learning Python.

---

**Mr. Mine can be found at:**
+ [Website](https://mrmine.com/game/ "Play on website")
+ [Steam](https://store.steampowered.com/app/1397920/MrMine/ "Play on Steam")

CHANGELOG.md
# **Changelog**
All changes to this project will be documented in this file.

*Date format: [dd-mm-year]*

Latest version: *1.0*

---

### **Version 1.0** - *[03/06-2022]*
* Created BOT & GUI
---

main.py
"""
MrMine BOT v1.0

A small program written in Python for the game named Mr. Mine.

Functions:
- Sell all your ores
- Forge gem [Missing yellow gem]

The program is made for a 1920x1080 resolution
"""

# TO-DO:
# 1) Create info text for user (What is the bot currently doing...)
# 2) Create info text for user (Total amount of ores sold etc)
# - Start & Stop button on one side, and 2x info text on other side.
# - BOT Info Text = Scroll Text
# 3) Create popup box for when activating all functions at once
# - Add function to forge yellow gem

from functools import partial
from kivy.app import App
from kivy.uix.widget import Widget
from kivy.properties import ObjectProperty
from kivy.clock import Clock
from kivy.core.window import Window
from sell_ores import sell_all_ores
from sell_ores import sell_all_ores_count
from forge_gems import forge_gem


class MainWidget(Widget):

    # Object properties [Used for linking with .kv file]
    btn_sell_all_ores = ObjectProperty(None)        # Sell all ores
    btn_forge_gem = ObjectProperty(None)            # Forge gem
    btn_forge_gem_color = ObjectProperty(None)      # Choose gem color
    btn_start_bot = ObjectProperty(None)            # Start BOT
    btn_stop_bot = ObjectProperty(None)             # Stop BOT
    label_info_for_user = ObjectProperty(None)      # Label for information

    sell_all_ores_timer = 60*30    # Timer for capacity to reach 100% (In seconds) | Set to 30 min

    # Time to forge a gem (In seconds)        Default forge time:
    red_gem_forge_time = 60*10                # 10 min
    blue_gem_forge_time = 60*30               # 30 min
    green_gem_forge_time = (60*60) * 4        # 4 hours
    purple_gem_forge_time = (60*60) * 16      # 16 hours
    yellow_gem_forge_time = (60*60) * 48      # 48 hours


    # Used for getting keyboard input
    def __init__(self, **kwargs):
        super(MainWidget, self).__init__(**kwargs)
        self._keyboard = Window.request_keyboard(self._keyboard_closed, self)
        self._keyboard.bind(on_key_down=self._on_keyboard_down)

    def _keyboard_closed(self):
        self._keyboard.unbind(on_key_down=self._on_keyboard_down)
        self._keyboard = None

    # If "Numpad 0" is pressed, stop bot
    def _on_keyboard_down(self, keyboard, keycode, text, modifiers):
        if keycode[1] == "numpad0":
            self.btn_click_stop_bot()

    # Button: Forge gem
    def btn_click_forge_gem(self):
        # Reset text for gem-color button
        if self.btn_forge_gem.state == "normal":
            self.btn_forge_gem_color.text = "Gem color?"

    # Button: Start BOT
    def btn_click_start_bot(self):
        # Switch activation state for buttons
        self.btn_start_bot.disabled = True
        self.btn_stop_bot.disabled = False
        self.btn_sell_all_ores.disabled = True
        self.btn_forge_gem.disabled = True
        self.btn_forge_gem_color.disabled = True

        # Check if "Sell all ores" is selected
        if self.btn_sell_all_ores.state == "down" and self.btn_forge_gem.state == "normal":
            print("Sell all ores\t Active")
            self.func_sell_all_ores = Clock.schedule_interval(sell_all_ores, self.sell_all_ores_timer) # Start function using Kivy Clock Module
            self.label_info_for_user.text = "All ores were sold " + str(sell_all_ores_count) + " times"

        elif self.btn_sell_all_ores.state == "normal":
            print("Sell all ores\t Deactive")

        # Check if "Forge gem" is selected
        if self.btn_forge_gem.state == "down" and self.btn_sell_all_ores.state == "normal":
            if self.btn_forge_gem_color.text == "Red":
                print("Forge red gem\t Active")
                self.func_forge_gem = Clock.schedule_interval(partial(forge_gem, "red"), self.red_gem_forge_time)
            if self.btn_forge_gem_color.text == "Blue":
                print("Forge blue gem\t Active")
                self.func_forge_gem = Clock.schedule_interval(partial(forge_gem, "blue"), self.blue_gem_forge_time)
            if self.btn_forge_gem_color.text == "Green":
                print("Forge green gem\t Active")
                self.func_forge_gem = Clock.schedule_interval(partial(forge_gem, "green"), self.green_gem_forge_time)
            if self.btn_forge_gem_color.text == "Purple":
                print("Forge purple gem\t Active")
                self.func_forge_gem = Clock.schedule_interval(partial(forge_gem, "purple"), self.purple_gem_forge_time)
            if self.btn_forge_gem_color.text == "Yellow":
                print("Forge yellow gem\t Active")
            if self.btn_forge_gem_color.text == "Gem color?":
                print("Forge gem\t Choose a gem color!")
        elif self.btn_forge_gem.state == "normal":
            print("Forge gem\t Deactive")

        # Check if everything is selected [W.I.P]
        if self.btn_sell_all_ores.state == "down" and self.btn_forge_gem.state == "down":
            print("\nThis function is not available at the moment.")

    # Button: Stop BOT
    def btn_click_stop_bot(self):
        # Switch activation state for buttons
        self.btn_start_bot.disabled = False
        self.btn_stop_bot.disabled = True
        self.btn_sell_all_ores.disabled = False
        self.btn_forge_gem.disabled = False
        self.btn_forge_gem_color.disabled = False

        # Stop active functions
        if self.btn_sell_all_ores.state == "down":
            self.func_sell_all_ores.cancel()
            self.btn_sell_all_ores.state = "normal"
            self.label_info_for_user.text = ""
        if self.btn_forge_gem.state == "down" and self.btn_forge_gem_color.text != "Gem color?" and self.btn_forge_gem_color.text != "Yellow": # W.I.P
            self.func_forge_gem.cancel()
            self.btn_forge_gem.state = "normal"


class MainApp(App):
    def build(self):
        self.title = "MrMine BOT v1.0" # Set title bar

MainApp().run() # Run app

forge_gems.py
"""
Program is currently set to ONLY look after Forge Station Lv. 3
Change "img_forge_station_lv3" to preferred level in if-statement (Line 92 + 95)
"""

import pyautogui
import time
import CustomFunctions as func # Custom module containing custom functions


# Image location paths
img_underground_city = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\underground_city.png"          #region=(1795,260,66,400)
img_forge_station = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_station.png"                #region=(170,450,500,450)
img_forge_station_lv2 = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_station_lv2.png"        #region=(170,450,500,450)
img_forge_station_lv3 = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_station_lv3.png"        #region=(170,450,500,450)
img_forge_red_gem = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_red_gem.png"                #region=(620,584,200,50)
img_forge_blue_gem = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_blue_gem.png"              #region=(860,584,200,50)
img_forge_green_gem = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_green_gem.png"            #region=(1100,584,200,50)
img_forge_purple_gem = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\forge_purple_gem.png"          #region=(620,730,200,50)
img_forge_yellow_gem = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\exit_btn.png"                  #region=()
img_exit_btn = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\underground_city\\exit_btn.png"                          #region=(1376,190,60,60)

# Time to forge a gem (In seconds)        Default forge time:
red_gem_forge_time = 60*10                # 10 min
blue_gem_forge_time = 60*30               # 30 min
green_gem_forge_time = (60*60) * 4        # 4 hours
purple_gem_forge_time = (60*60) * 16      # 16 hours
yellow_gem_forge_time = (60*60) * 48      # 48 hours

# Total gem-slots in Forge Station
red_gem_slots = 12
blue_gem_slots = 6
green_gem_slots = 4
purple_gem_slots = 3

total_forged_gems = 0 # Counter for total forged gems

# States for vissibility for Underground City & inventory of Forge Station
undergroud_city_is_vissible = False
forge_inventory_is_vissible = False

# Reset vissibility state of "items" [Kind of a reset to "forge_gem" function]
def reset_vissible_states():
    global undergroud_city_is_vissible
    global forge_inventory_is_vissible 
    undergroud_city_is_vissible = False
    forge_inventory_is_vissible = False

# Exit Forge Station inventory
def exit_forge_inventory():

    global undergroud_city_is_vissible
    global forge_inventory_is_vissible

    # Find and click "Exit" button
    func.msg_time("Locating \"Exit-button\"...")
    if pyautogui.locateOnScreen(img_exit_btn, region=(1376,190,60,60), grayscale=True, confidence=0.8) != None:
        func.msg_time("\"Exit-button\" located!")
        x_exit_btn, y_exit_btn = pyautogui.locateCenterOnScreen(img_exit_btn, region=(1376,190,60,60), grayscale=True, confidence=0.8)
        func.click(x_exit_btn, y_exit_btn)
        forge_inventory_is_vissible = False
    else:
        reset_vissible_states()
        func.msg_time("Unable to locate \"Exit-button\"...")
    
    time.sleep(0.1)

# Forge gem depending on given color
def forge_gem(gem_color="", *args):
    
    global undergroud_city_is_vissible
    global forge_inventory_is_vissible
    global total_forged_gems # Total forged gems

    # ----------| UNDERGROUD CITY |---------- #
    if undergroud_city_is_vissible == False: # If Underground City is not vissible, find and click it
        func.msg_time("Locating Underground City...") # Print message
        if pyautogui.locateOnScreen(img_underground_city, region=(1795,260,66,400), grayscale=True, confidence=0.8) != None: # Locate "Underground City" button
            func.msg_time("Underground City located!")
            x_undergroud_city, y_underground_city = pyautogui.locateCenterOnScreen(img_underground_city, region=(1795,260,66,400), grayscale=True, confidence=0.8) # Create x and y for center location of button
            func.click(x_undergroud_city, y_underground_city) # Click the "Underground City" button
            undergroud_city_is_vissible = True # Set vissible state of Underground City to TRUE
        else: # If not possible to find "Underground City" button, let user know (Kind of error-message)
            reset_vissible_states() # Reset vissibility states of "items"
            func.msg_time("Unable to locate Underground City...")
    
    time.sleep(0.1)

    # ----------| FORGE STATION |---------- #
    if forge_inventory_is_vissible == False and undergroud_city_is_vissible == True:
        func.msg_time("Locating Forge Station...")
        if pyautogui.locateOnScreen(img_forge_station_lv3, region=(170,450,500,450), grayscale=True, confidence=0.35) != None: # Confidence needs to be low due to in-game animations 
            func.msg_time("Forge Station located!")
            x_forge_station, y_forge_station = pyautogui.locateCenterOnScreen(img_forge_station_lv3, region=(170,450,500,450), grayscale=True, confidence=0.35)
            func.click(x_forge_station, y_forge_station)
            forge_inventory_is_vissible = True
        else:
            reset_vissible_states()
            func.msg_time("Unable to locate Forge Station...")
    
    time.sleep(0.1)

    # ----------| INVENTORY OF FORGE STATION |---------- #
    if forge_inventory_is_vissible == True:
        func.msg_time("Locating gem to forge...")
        
        # Check what color gem to forge:
        # ----------| RED GEM |---------- #
        if gem_color == "red": # Red gem
            if pyautogui.locateOnScreen(img_forge_red_gem, region=(620,584,200,50), grayscale=True, confidence=0.9) != None:
                x_forge_red_gem, y_forge_red_gem = pyautogui.locateCenterOnScreen(img_forge_red_gem, region=(620,584,200,50), grayscale=True, confidence=0.9)
                func.click(x_forge_red_gem, y_forge_red_gem, clicks=red_gem_slots)
                total_forged_gems += red_gem_slots # Increase total amount of forged gems
                red_gem_msg = str(red_gem_slots) + " red gems were forged\t" + str(total_forged_gems) + " gems were forged in total" # Create message for user
                func.msg_time(red_gem_msg) # Outprint message to user
                time.sleep(0.1)
                exit_forge_inventory() # Exit inventory
            else:
                reset_vissible_states()
                func.msg_time("Unable to locate red gem...")
 
        # ----------| BLUE GEM |---------- #
        elif gem_color == "blue": # Blue gem
            if pyautogui.locateOnScreen(img_forge_blue_gem, region=(860,584,200,50), grayscale=True, confidence=0.9) != None:
                x_forge_blue_gem, y_forge_blue_gem = pyautogui.locateCenterOnScreen(img_forge_blue_gem, region=(860,584,200,50), grayscale=True, confidence=0.9)
                func.click(x_forge_blue_gem, y_forge_blue_gem, clicks=blue_gem_slots)
                total_forged_gems += blue_gem_slots
                blue_gem_msg = str(blue_gem_slots) + " blue gems were forged\t" + str(total_forged_gems) + " gems were forged in total"
                func.msg_time(blue_gem_msg)
                time.sleep(0.1) 
                exit_forge_inventory()
            else:
                reset_vissible_states()
                func.msg_time("Unable to blue locate gem...")

        # ----------| GREEN GEM |---------- #
        elif gem_color == "green": # Green gem
            if pyautogui.locateOnScreen(img_forge_green_gem, region=(1100,584,200,50), grayscale=True, confidence=0.9) != None:
                x_forge_green_gem, y_forge_green_gem = pyautogui.locateCenterOnScreen(img_forge_green_gem, region=(1100,584,200,50), grayscale=True, confidence=0.9)
                func.click(x_forge_green_gem, y_forge_green_gem, clicks=green_gem_slots)
                total_forged_gems += green_gem_slots
                green_gem_msg = str(green_gem_slots) + " green gems were forged\t" + str(total_forged_gems) + " gems were forged in total"
                func.msg_time(green_gem_msg)
                time.sleep(0.1)
                exit_forge_inventory()
            else:
                reset_vissible_states()
                func.msg_time("Unable to locate green gem...")

        # ----------| PURPLE GEM |---------- #
        elif gem_color == "purple": # Purple gem
            if pyautogui.locateOnScreen(img_forge_purple_gem, region=(620,730,200,50), grayscale=True, confidence=0.9) != None:
                x_forge_purple_gem, y_forge_purple_gem = pyautogui.locateCenterOnScreen(img_forge_purple_gem, region=(620,730,200,50), grayscale=True, confidence=0.9)
                func.click(x_forge_purple_gem, y_forge_purple_gem, clicks=purple_gem_slots)
                total_forged_gems += purple_gem_slots
                purple_gem_msg = str(purple_gem_slots) + " purple gems were forged\t" + str(total_forged_gems) + " gems were forged in total"
                func.msg_time(purple_gem_msg)
                time.sleep(0.1)
                exit_forge_inventory()
            else:
                reset_vissible_states()
                func.msg_time("Unable to locate purple gem...")
    
    time.sleep(0.1)


CustomFunctions.py
"""
A file containing randoms functions that are used in the code

Instead of copy'n'paste the same functions in each file, I just
have one main file - Makes it easiere if you want to edit the
functions.
"""


import time
import win32api, win32con


# Left-click for mouse
def click(x,y, clicks=1):

    for click in range(0, clicks): # Left-click X amount of times
        win32api.SetCursorPos((x,y)) # Set mouse-cursor position
        time.sleep(0.01) # Wait 0.01 second
        win32api.mouse_event(win32con.MOUSEEVENTF_LEFTDOWN,0,0) # Press left click
        time.sleep(0.01) # Wait 0.01 second
        win32api.mouse_event(win32con.MOUSEEVENTF_LEFTUP,0,0) # Release left click
        click += 1 # Increase total clicks
        time.sleep(0.1) # Wait 0.1 second

    time.sleep(0.5) # Wait 0.5 second. (Needed for click function to work properply in-game)

# Print time-stamp
def msg_time(text=""):
    print("[" + time.ctime()[11:19] + "] " + str(text))

sell_ores.py
import pyautogui
import time
import CustomFunctions as func # Custom module containing custom functions


sell_all_ores_count = 0        # Counter for selling all ores
full_capacity_timer = 60*25    # Timer for capacity to reach 100% (In seconds)

# Image location paths
img_sell_center = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\sell_center\\sell_center.png"      #region=(330,480,300,300)
img_sell_all_ores = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\sell_center\\sell_all_ores.png"  #region=(500,730,230,60)
img_exit_btn = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\sell_center\\exit_btn.png"            #region=(1376,250,60,60)
img_cave_building = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\sell_center\\cave_building.png"  #region=(1795,260,66,400)
img_move_to_top = "C:\\Users\\Victor\\Documents\\Visual Studio Code - Workplace\\MrMineBOT\\img\\sell_center\\move_to_top.png"      #region=(64,144,90,90)

# State for vissibility for Sell Center
sell_center_inventory_is_vissible = False # Default = False


# Exit Forge Station inventory
def exit_sell_center_inventory():

    global sell_center_inventory_is_vissible

    # Find and click "Exit" button
    func.msg_time("Locating \"Exit-button\"...")
    if pyautogui.locateOnScreen(img_exit_btn, region=(1376,250,60,60), grayscale=True, confidence=0.75) != None:
        func.msg_time("\"Exit-button\" located!")
        x_exit_btn, y_exit_btn = pyautogui.locateCenterOnScreen(img_exit_btn, region=(1376,250,60,60), grayscale=True, confidence=0.75)
        func.click(x_exit_btn, y_exit_btn)
        sell_center_inventory_is_vissible = False
    else:
        sell_center_inventory_is_vissible = False
        func.msg_time("Unable to locate \"Exit-button\"...")
    
    time.sleep(0.1)

# Sell all ores at "Sell Center"
def sell_all_ores(*args):
    
    global sell_all_ores_count # Let function use "sell_all_ores_count" variable
    global sell_center_inventory_is_vissible

    if sell_center_inventory_is_vissible == False: # Check if Sell Center is vissible
        func.msg_time("Locating Sell Center...")

        # ----------| SELL CENTER |---------- #
        if pyautogui.locateOnScreen(img_sell_center, region=(330,480,300,300), grayscale=True, confidence=0.8) != None: # Press Sell Center, if vissible
            func.msg_time("Sell Center located!")
            x_sell_center, y_sell_center = pyautogui.locateCenterOnScreen(img_sell_center, region=(330,480,300,300), grayscale=True, confidence=0.8) # Assign x- & y-values of "Sell Center"
            func.click(x_sell_center, y_sell_center) # Click on Sell Center
            sell_center_inventory_is_vissible = True
        else: # If not vissible, press cave building & move to top
            sell_center_inventory_is_vissible = False
            func.msg_time("Unable to locate Sell Center...")

            time.sleep(0.1)

            # ----------| CAVE BUILDING|---------- #
            func.msg_time("Locating Cave Building...")
            if pyautogui.locateOnScreen(img_cave_building, region=(1795,260,66,400), grayscale=True, confidence=0.8) != None:
                func.msg_time("Cave Building located!")
                x_cave_building, y_cave_building = pyautogui.locateCenterOnScreen(img_cave_building, region=(1795,260,66,400), grayscale=True, confidence=0.8)
                func.click(x_cave_building, y_cave_building)
            else:
                func.msg_time("Unable to locate Cave Building...")

            time.sleep(0.1)

            # ----------| "GO-TO-TOP" BUTTON |---------- #
            func.msg_time("Locating \"Go-to-top\" button...")
            if pyautogui.locateOnScreen(img_move_to_top, region=(64,144,90,90), grayscale=True, confidence=0.8) != None: # Go to top
                func.msg_time("\"Go-to-top\" button located!")
                x_move_to_top, y_move_to_top = pyautogui.locateCenterOnScreen(img_move_to_top, region=(64,144,90,90), grayscale=True, confidence=0.8)
                func.click(x_move_to_top, y_move_to_top)
            else:
                func.msg_time("Unable to locate \"Go-to-top\" button...")
  
    time.sleep(0.1)

    # ----------| INVENTORY OF SELL CENTER |---------- #
    if sell_center_inventory_is_vissible == True:
        func.msg_time("Locating \"Sell all\" button")

        if pyautogui.locateOnScreen(img_sell_all_ores, region=(500,730,230,60), grayscale=True, confidence=0.8) != None:
            func.msg_time("\"Sell all\" button located!")
            x_sell_all_ores, y_sell_all_ores = pyautogui.locateCenterOnScreen(img_sell_all_ores, region=(500,730,230,60), grayscale=True, confidence=0.8)
            func.click(x_sell_all_ores, y_sell_all_ores)
            sell_all_ores_count += 1 # Increase counter for selling all ores
            sell_all_ores_msg = "All ores were sold " + str(sell_all_ores_count) + " times" # Create message for user
            func.msg_time(sell_all_ores_msg) # Print message
            exit_sell_center_inventory()
        else:
            sell_center_inventory_is_vissible = False
            func.msg_time("Unable to locate \"Sell all\" button...")
            
    time.sleep(0.1)
